### PR TITLE
Kulikov Artem. Lab 3. Var 1

### PR DESCRIPTION
**Pull request by AveZorgen**

### CODE REVIEW

**Comment by mahbhlddnhakkh** [comment]
I think you should add a patch that allows to not modify `llvm/lib/Target/X86/CMakeLists.txt`, `llvm/lib/Target/X86/X86.h` and `llvm/lib/Target/X86/X86TargetMachine.cpp` for every lab so it will add/register new labs automatically. Or else it's a recipe for conflicts disaster.

For CMakeLists.txt there should be solutions already. As for `llvm/lib/Target/X86/X86.h` and `llvm/lib/Target/X86/X86TargetMachine.cpp`: preprocessor is an option. And to grab lab's headers files, you can try Nesterov's approach. [File](https://github.com/learning-process/ppc-2024-threads/blob/master/tasks/CMakeLists.txt#L33) . `add_compile_definitions(PATH_TO_PPC_PROJECT="${CMAKE_SOURCE_DIR}")`. In C++ code we can use PATH_TO_PPC_PROJECT as a string path to project root folder. You need to do something similar. E.g. store to constant list of lab's specific files or something like that and in `llvm/lib/Target/X86/X86.h` and `llvm/lib/Target/X86/X86TargetMachine.cpp` somehow include required functions with preprocessor.

**Comment by nol0n** [comment]
@mahbhlddnhakkh I think you could try to implement it yourself and offer your solution.

**Comment by mahbhlddnhakkh** [comment]
Also the pass is not in the `llvm/labs/lab3` directory (but it should be), therefore I can't use this lab as a basis for the conflicts solution and I still haven't looked at the code, so... Nuh-uh.

**Comment by mahbhlddnhakkh** [comment]
I came up with one solution. [Something like this should be implemented.](https://cloud.unn.ru/s/sNHXbCzDzZGExoW) And C++ code for passes should be in `llvm/lib/Target/X86/lab3` directory. +function names should follow the same template as this lab.
`<labname>.cpp`
`FunctionPass *create<labname>Pass();`
`void initialize<labname>PassPass(PassRegistry &);`

**Comment by AveZorgen** [comment]
@m-ly4

File: llvm/test/lab3/kulikov_artem/vfmadd.mir [hunk 1]
```diff
@@ -0,0 +1,497 @@
+# RUN: llc -mtriple x86_64-unknown-linux-gnu -run-pass=x86-kulikov-fma  %s -o - | FileCheck %s
+
+# src.c
```
Comment (m-ly4) [inline llvm/test/lab3/kulikov_artem/vfmadd.mir:3]: What output will be for the following example?
```
# __m128d foo(__m128d a, __m128d b, __m128d c, __m128d d) { 
 #     __m128d tmp = a * b; 
 #     c = tmp + b;
 #     d = tmp + c; 
 #     return d;
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/kulikov_artem/vfmadd.mir (hunk 1 above)
Comment (AveZorgen) [reply llvm/test/lab3/kulikov_artem/vfmadd.mir:3]: %1:vr128 = COPY $xmm1
%0:vr128 = COPY $xmm0
%4:vr128 = nofpexcept MULPDrr %0, %1, implicit $mxcsr
%5:vr128 = nofpexcept ADDPDrr %4, %1, implicit $mxcsr
%6:vr128 = nofpexcept ADDPDrr %4, killed %5, implicit $mxcsr
$xmm0 = COPY %6
RET 0, $xmm0

**Review by m-ly4 — APPROVED** [review]

File: llvm/lib/Target/X86/CMakeLists.txt [hunk 2]
```diff
@@ -118,3 +118,4 @@ add_subdirectory(Disassembler)
 add_subdirectory(MCA)
 add_subdirectory(MCTargetDesc)
 add_subdirectory(TargetInfo)
+add_subdirectory(labs)
```
Comment (m-ly4) [inline llvm/lib/Target/X86/CMakeLists.txt]:
```suggestion
add_subdirectory(lab3)
```

File: llvm/lib/Target/X86/labs/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+if (NOT WIN32 AND NOT CYGWIN)
```
Comment (m-ly4) [inline llvm/lib/Target/X86/labs/CMakeLists.txt]: Please rename dir to lab3

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Review by m-ly4 — APPROVED** [review]

File: llvm/lib/Target/X86/lab3/kulikov_artem/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+file(GLOB_RECURSE ALL_SOURCE_FILES *.cpp *.h)
+add_llvm_pass_plugin(${PluginName} 
+  ${ALL_SOURCE_FILES}     
+  DEPENDS
+  intrinsics_gen
+  X86
+  BUILDTREE_ONLY)
+
+target_include_directories(${PluginName} PUBLIC ${PATH_TO_X86})
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/kulikov_artem/CMakeLists.txt:11]:
```suggestion
target_include_directories(${PluginName} PUBLIC ${PATH_TO_X86})

set(LLVM_TEST_DEPENDS ${PluginName} ${LLVM_TEST_DEPENDS} PARENT_SCOPE)
```

**Pull request merged**